### PR TITLE
Add NPU device type and three fp16 models for image classification

### PR DESCRIPTION
**Pull request by mingmingtasd**
PTAL, thanks! @Honry @huningxin @ibelem

### CODE REVIEW

**Comment by mingmingtasd** [comment]
Need we show info logs on page like: efficientnet model requires DirectML 1.14?

**Comment by huningxin** [comment]
I think we can extend [WebNN Installation Guides](https://github.com/webmachinelearning/webnn-samples?tab=readme-ov-file#webnn-installation-guides) to help developers test WebNN on NPU. For example, through [DirectML developer preview for NPU on new Intel® Core™ Ultra devices](https://devblogs.microsoft.com/directx/introducing-neural-processor-unit-npu-support-in-directml-developer-preview/) on Windows. We can extend the guide for other platforms later.

File: image_classification/main.js [hunk 1]
```diff
@@ -35,13 +39,68 @@ let lastBackend = '';
...
+    ],
+  },
+  'npu': {
+    'float16': [
+      'efficientnet',
+      'mobilenet',
+      'resnet50v1',
+    ],
+  },
+};
```
Comment (Honry) [inline image_classification/main.js:76]: You should also rename the resnet50 in these two constants.

File: image_classification/index.html [hunk 2]
```diff
@@ -43,6 +43,9 @@
               <label class="btn btn-outline-info custom" name="webnn">
                 <input type="radio" name="backend" id="webnn_gpu" autocomplete="off">WebNN (GPU)
               </label>
+              <label class="btn btn-outline-info custom" name="webnn">
+                <input type="radio" name="backend" id="webnn_npu" autocomplete="off">WebNN (NPU)
```
Comment (Honry) [inline image_classification/index.html:47]: Shall we disable NPU if the test browser doesn't support it? Similar to the logic of disabling the `Wasm (CPU)`.

Comment (mingmingtasd) [reply image_classification/index.html:47]: Good idea, done.

**Comment by Honry** [comment]
cc/ [@JiangYueBook](https://github.com/JiangYueBook), this change would impact your automation tool.

**Comment by mingmingtasd** [comment]
> I think we can extend [WebNN Installation Guides](https://github.com/webmachinelearning/webnn-samples?tab=readme-ov-file#webnn-installation-guides) to help developers test WebNN on NPU. For example, through [DirectML developer preview for NPU on new Intel® Core™ Ultra devices](https://devblogs.microsoft.com/directx/introducing-neural-processor-unit-npu-support-in-directml-developer-preview/) on Windows. We can extend the guide for other platforms later.

Yes, we should add, can I send another PR for it later? @huningxin

File: common/utils.js [hunk 3]
```diff
@@ -188,11 +297,12 @@ export function getInputTensor(inputElement, inputOptions) {
           value = pixels[h * width * imageChannels + w * imageChannels + c];
         }
         if (inputLayout === 'nchw') {
-          tensor[c * width * height + h * width + w] =
-              (value - mean[c]) / std[c];
+          tensor[c * width * height + h * width + w] = dataType === 'float16' ?
+              toHalf((value - mean[c]) / std[c]) : (value - mean[c]) / std[c];
```
Comment (huningxin) [inline common/utils.js:301]: Converting float32 to float16 in JS is slow, try WebNN `cast` operator.

File: image_classification/main.js [hunk 4]
```diff
@@ -154,7 +248,31 @@ async function renderCamStream() {
...
+  // NPU dosen't support softmax.
+  // Convert output buffer from float16 to float32, because
+  // tf.tensor/tf.softmax doesn't support float16 data type
+  // according to https://js.tensorflow.org/api/latest/#tensor.
+  // TODO: Remove this workaround once NPU supports softmax.
+  if (inputOptions.dataType === 'float16') {
+    const elementsCount = utils.sizeOfShape(netInstance.outputDimensions);
+    const float32Array = new Float32Array(elementsCount);
+    for (let i = 0; i < elementsCount; ++i) {
+      float32Array[i] = utils.float16ToNumber(buffer[i]);
```
Comment (huningxin) [inline image_classification/main.js:262]: Converting float32 to float16 is slow, please try to use WebNN cast operator.

File: common/utils.js (hunk 3 above)
Comment (mingmingtasd) [reply common/utils.js]: Done, thanks!

File: image_classification/main.js (hunk 4 above)
Comment (mingmingtasd) [reply image_classification/main.js]: Done, thanks!

File: image_classification/efficientnet_fp16_nchw.js [hunk 5]
```diff
@@ -17,7 +18,7 @@ export class EfficientNetFP16Nchw {
       inputLayout: 'nchw',
       labelUrl: './labels/labels1000.txt',
       inputDimensions: [1, 3, 224, 224],
-      dataType: 'float16',
+      dataType: 'float32',
```
Comment (Honry) [inline image_classification/efficientnet_fp16_nchw.js:21]: Ditto.

File: image_classification/efficientnet_fp16_nchw.js [hunk 6]
```diff
@@ -8,6 +8,7 @@ export class EfficientNetFP16Nchw {
     this.context_ = null;
     this.builder_ = null;
     this.graph_ = null;
+    this.targetDataType_ = 'float16';
```
Comment (Honry) [inline image_classification/efficientnet_fp16_nchw.js:11]: I think you don't need this variable, as this model is fp16 only.

File: image_classification/efficientnet_fp16_nchw.js [hunk 7]
```diff
@@ -30,8 +31,10 @@ export class EfficientNetFP16Nchw {
     } else {
       prefix = this.weightsUrl_ + 'conv' + name;
     }
-    const weight = buildConstantByNpy(this.builder_, prefix + '_w.npy');
-    options.bias = await buildConstantByNpy(this.builder_, prefix + '_b.npy');
+    const weight = buildConstantByNpy(this.builder_, prefix + '_w.npy',
+        this.targetDataType_ = 'float16');
```
Comment (Honry) [inline image_classification/efficientnet_fp16_nchw.js:35]: You can pass 'float16' directly.

File: image_classification/efficientnet_fp16_nchw.js [hunk 8]
```diff
@@ -142,7 +148,13 @@ export class EfficientNetFP16Nchw {
     const conv22 = this.buildConv_(conv21, '0', '', true);
     const pool1 = this.builder_.averagePool2d(await conv22);
     const reshape = this.builder_.reshape(pool1, [1, 1280]);
-    return this.buildGemm_(reshape, '0');
+    const gemm = this.buildGemm_(reshape, '0');
+    if (contextOptions.deviceType === 'npu') {
+      return this.builder_.cast(await gemm, 'float32');
```
Comment (Honry) [inline image_classification/efficientnet_fp16_nchw.js:153]: I suppose you don't need `await` for `gemm`.

File: image_classification/efficientnet_fp16_nchw.js [hunk 9]
```diff
@@ -142,7 +148,13 @@ export class EfficientNetFP16Nchw {
...
+    } else {
+      const softmax = this.builder_.softmax(await gemm);
```
Comment (Honry) [inline image_classification/efficientnet_fp16_nchw.js:155]: Ditto.

File: image_classification/mobilenet_nchw.js [hunk 10]
```diff
@@ -4,19 +4,20 @@ import {buildConstantByNpy, weightsOrigin} from '../common/utils.js';
 
 // MobileNet V2 model with 'nchw' input layout
 export class MobileNetV2Nchw {
-  constructor(dataType = 'float32') {
+  constructor(targetDataType = 'float32') {
     this.context_ = null;
     this.deviceType_ = null;
     this.builder_ = null;
     this.graph_ = null;
+    this.targetDataType_ = targetDataType;
```
Comment (Honry) [inline image_classification/mobilenet_nchw.js:12]: Ditto, you don't need to use this  `this.targetDataType_`, your previous code should work.

File: image_classification/resnet50v1_fp16_nchw.js [hunk 11]
```diff
@@ -8,6 +8,7 @@ export class ResNet50V1FP16Nchw {
     this.context_ = null;
     this.builder_ = null;
     this.graph_ = null;
+    this.targetDataType_ = 'float16';
```
Comment (Honry) [inline image_classification/resnet50v1_fp16_nchw.js:11]: Ditto.

**Comment by mingmingtasd** [comment]
PTAL, thanks! @Honry

File: image_classification/mobilenet_nchw.js [hunk 12]
```diff
@@ -26,25 +26,24 @@ export class MobileNetV2Nchw {
...
     };
     this.outputDimensions = [1, 1000];
   }
 
   async buildConv_(input, name, relu6 = true, options = {}) {
     let weights;
-    if (this.targetDataType_==='float32') {
+    if (this.dataType_ === 'float32') {
       weights = buildConstantByNpy(this.builder_,
-          `${this.weightsUrl_}conv_${name}_weight.npy`, this.targetDataType_);
```
Comment (Honry) [inline image_classification/mobilenet_nchw.js]: You can simplify this code by passing `this.dataType_` to `buildConstantByNpy` directly.

**Review by Honry — APPROVED** [review]
LGTM, thanks! @huningxin, would you like to take another look?

**Review by huningxin — APPROVED** [review]
LGTM

**Pull request merged**